src/AzureExtension/DevBox/DevBoxInstance.cs

```diff
@@ -685,116 +717,138 @@ public IAsyncOperation<ComputeSystemOperationResult> DoPinActionAsync(string loc
     }
 
     public IAsyncOperation<ComputeSystemOperationResult> PinToStartMenuAsync()
     {
+        // Since there is no 'Pinning' state we'll say that the state is saving
+        StateChanged?.Invoke(this, ComputeSystemState.Saving);
         return DoPinActionAsync("startMenu", "pin");
     }
 
     public IAsyncOperation<ComputeSystemOperationResult> UnpinFromStartMenuAsync()
     {
+        // Since there is no 'Unpinning' state we'll say that the state is saving
+        StateChanged?.Invoke(this, ComputeSystemState.Saving);
         return DoPinActionAsync("startMenu", "unpin");
     }
 
     public IAsyncOperation<ComputeSystemOperationResult> PinToTaskbarAsync()
     {
+        // Since there is no 'Pinning' state we'll say that the state is saving
+        StateChanged?.Invoke(this, ComputeSystemState.Saving);
         return DoPinActionAsync("taskbar", "pin");
     }
 
     public IAsyncOperation<ComputeSystemOperationResult> UnpinFromTaskbarAsync()
     {
+        // Since there is no 'Unpinning' state we'll say that the state is saving
+        StateChanged?.Invoke(this, ComputeSystemState.Saving);
         return DoPinActionAsync("taskbar", "unpin");
     }
 
-    public IAsyncOperation<ComputeSystemPinnedResult> GetPinStatusAsync(string location)
+    public IAsyncOperation<ComputeSystemPinnedResult> GetPinStatusAsync(string location, string errorMessage)
     {
         return Task.Run(() =>
         {
-            var validationString = ValidateWindowsAppParameters();
-            if (!string.IsNullOrEmpty(validationString))
-            {
-                _log.Error(validationString);
-                return new ComputeSystemPinnedResult(new NotSupportedException(), Resources.GetResource(Constants.DevBoxUnableToPerformOperationKey), validationString);
-            }
-
-            var exitcode = ExitCodeInvalid;
-            var psi = new ProcessStartInfo();
-            psi.UseShellExecute = true;
-            psi.FileName = string.Format(CultureInfo.InvariantCulture, ProtocolPinString, location, "status", WorkspaceId, DisplayName, Environment, Username);
-            Process? p = Process.Start(psi);
-            if (p != null)
+            try
             {
-                p.Refresh();
-                AllowSetForegroundWindow(p.Id);
-
-                // This signals to the WindowsApp that it has been given foreground rights
-                EventWaitHandle signalForegroundSet = new EventWaitHandle(false, EventResetMode.AutoReset, WindowsAppEventName);
-                signalForegroundSet.Set();
-
-                p.WaitForExit();
-                exitcode = p.ExitCode;
-                if (exitcode == ExitCodePinned)
+                var validationString = ValidateWindowsAppParameters();
+                if (!string.IsNullOrEmpty(validationString))
                 {
-                    return new ComputeSystemPinnedResult(true);
+                    throw new NotSupportedException(validationString);

```

Comment:
I thought we don't want to throw exceptions in the extension.

Reply:
I believe I wrapped this in a try/catch. Since we're doing a create process as well which can throw I just made the other error scenarios throw and catch them in the try/catch and return it back to Dev Home